Prevent default warnings for deprecated KV API in C++ API

Currently when using the C++ API at all we see deprecation warnings about the KV API even when not using it explicitly:

```
In file included from /Users/tyler/tiledb/TileDB/dist/include/tiledb/query.h:38:
In file included from /Users/tyler/tiledb/TileDB/dist/include/tiledb/array.h:38:
In file included from /Users/tyler/tiledb/TileDB/dist/include/tiledb/array_schema.h:38:
In file included from /Users/tyler/tiledb/TileDB/dist/include/tiledb/attribute.h:39:
/Users/tyler/tiledb/TileDB/dist/include/tiledb/deleter.h:98:5: warning: 'tiledb_kv_iter_free' is deprecated [-Wdeprecated-declarations]
    tiledb_kv_iter_free(&p);
    ^
/Users/tyler/tiledb/TileDB/dist/include/tiledb/tiledb.h:5119:15: note: 'tiledb_kv_iter_free' has been explicitly marked deprecated here
TILEDB_EXPORT TILEDB_DEPRECATED void tiledb_kv_iter_free(
```

This can be fixed by
- separating the KV deleter functions into a separate class
- removing the `map` and `map_schema` includes from the top-level `#include <tiledb/tiledb>` file.

Users who wish to use the KV API will have to include the `tiledb/map*` files explicitly, and then deal with the deprecation warnings.

--- a/HISTORY.md
+++ b/HISTORY.md
@@ -60,6 +60,7 @@
 * Removed cast operators of C++ API objects to their underlying C API objects. This helps prevent inadvertent memory issues such as double-frees.
 * Removed ability to set `null` tile extents on dimensions. All dimensions must now have an explicit tile extent.
 * Changed argument `config` in `Array::consolidate()` from a const-ref to a pointer.
+* Removed default includes of `Map` and `MapSchema`. To use the deprecated KV API temporarily, include `<tiledb/map.h>` explicitly.
 
 # TileDB v1.5.0 Release Notes
 

--- a/examples/cpp_api/map.cc
+++ b/examples/cpp_api/map.cc
@@ -35,6 +35,7 @@
  *
  */
 
+#include <tiledb/map.h>
 #include <iostream>
 #include <tiledb/tiledb>
 

--- a/examples/cpp_api/object.cc
+++ b/examples/cpp_api/object.cc
@@ -47,6 +47,7 @@
  * move/remove TileDB objects.
  */
 
+#include <tiledb/map.h>
 #include <iostream>
 #include <tiledb/tiledb>
 

--- a/examples/cpp_api/quickstart_map.cc
+++ b/examples/cpp_api/quickstart_map.cc
@@ -35,6 +35,7 @@
  *
  */
 
+#include <tiledb/map.h>
 #include <iostream>
 #include <tiledb/tiledb>
 

--- a/test/src/unit-cppapi-map.cc
+++ b/test/src/unit-cppapi-map.cc
@@ -31,6 +31,7 @@
  */
 
 #include "catch.hpp"
+#include "tiledb/sm/cpp_api/map.h"
 #include "tiledb/sm/cpp_api/tiledb"
 #include "tiledb/sm/misc/utils.h"
 

--- a/test/src/unit-cppapi-schema.cc
+++ b/test/src/unit-cppapi-schema.cc
@@ -31,6 +31,7 @@
  */
 
 #include "catch.hpp"
+#include "tiledb/sm/cpp_api/map.h"
 #include "tiledb/sm/cpp_api/tiledb"
 
 TEST_CASE("C++ API: Schema", "[cppapi]") {

--- a/tiledb/sm/cpp_api/deleter.h
+++ b/tiledb/sm/cpp_api/deleter.h
@@ -82,22 +82,6 @@ class Deleter {
     tiledb_array_schema_free(&p);
   }
 
-  void operator()(tiledb_kv_t* p) const {
-    tiledb_kv_free(&p);
-  }
-
-  void operator()(tiledb_kv_schema_t* p) const {
-    tiledb_kv_schema_free(&p);
-  }
-
-  void operator()(tiledb_kv_item_t* p) const {
-    tiledb_kv_item_free(&p);
-  }
-
-  void operator()(tiledb_kv_iter_t* p) const {
-    tiledb_kv_iter_free(&p);
-  }
-
   void operator()(tiledb_attribute_t* p) const {
     tiledb_attribute_free(&p);
   }

--- a/tiledb/sm/cpp_api/map.h
+++ b/tiledb/sm/cpp_api/map.h
@@ -286,7 +286,7 @@ class MapItem {
   std::shared_ptr<tiledb_kv_item_t> item_;
 
   /** Deleter for tiledb object **/
-  impl::Deleter deleter_;
+  impl::KVDeleter deleter_;
 
   /** Underlying Map **/
   Map* map_ = nullptr;
@@ -562,7 +562,7 @@ class MapIter : public std::iterator<std::forward_iterator_tag, MapItem> {
  private:
   /** Base map. **/
   Map* map_;
-  impl::Deleter deleter_;
+  impl::KVDeleter deleter_;
 
   /** Current item **/
   std::unique_ptr<MapItem> item_;
@@ -1520,7 +1520,7 @@ class Map {
   std::shared_ptr<tiledb_kv_t> kv_;
 
   /** Closes the map on destruction. **/
-  impl::Deleter deleter_;
+  impl::KVDeleter deleter_;
 
   /** URI **/
   std::string uri_;

--- a/tiledb/sm/cpp_api/map_schema.h
+++ b/tiledb/sm/cpp_api/map_schema.h
@@ -49,6 +49,42 @@
 
 namespace tiledb {
 
+namespace impl {
+/**
+ * Deleter for KV TileDB C types. Used to suppress deprecation warnings for
+ * the KV API unless explicitly used.
+ */
+class KVDeleter {
+ public:
+  /* ********************************* */
+  /*     CONSTRUCTORS & DESTRUCTORS    */
+  /* ********************************* */
+
+  KVDeleter() = default;
+  KVDeleter(const KVDeleter&) = default;
+
+  /* ********************************* */
+  /*              DELETERS             */
+  /* ********************************* */
+
+  void operator()(tiledb_kv_t* p) const {
+    tiledb_kv_free(&p);
+  }
+
+  void operator()(tiledb_kv_schema_t* p) const {
+    tiledb_kv_schema_free(&p);
+  }
+
+  void operator()(tiledb_kv_item_t* p) const {
+    tiledb_kv_item_free(&p);
+  }
+
+  void operator()(tiledb_kv_iter_t* p) const {
+    tiledb_kv_iter_free(&p);
+  }
+};
+}  // namespace impl
+
 /**
  * Represents the schema of a Map (key-value store).
  *
@@ -81,7 +117,7 @@ class MapSchema : public Schema {
       : Schema(ctx) {
     tiledb_kv_schema_t* schema;
     ctx.handle_error(tiledb_kv_schema_alloc(ctx.ptr().get(), &schema));
-    schema_ = std::shared_ptr<tiledb_kv_schema_t>(schema, deleter_);
+    schema_ = std::shared_ptr<tiledb_kv_schema_t>(schema, kv_deleter_);
   }
 
   /**
@@ -118,7 +154,7 @@ class MapSchema : public Schema {
         encryption_key,
         key_length,
         &schema));
-    schema_ = std::shared_ptr<tiledb_kv_schema_t>(schema, deleter_);
+    schema_ = std::shared_ptr<tiledb_kv_schema_t>(schema, kv_deleter_);
   }
 
   /**
@@ -148,7 +184,7 @@ class MapSchema : public Schema {
    */
   MapSchema(const Context& ctx, tiledb_kv_schema_t* schema)
       : Schema(ctx) {
-    schema_ = std::shared_ptr<tiledb_kv_schema_t>(schema, deleter_);
+    schema_ = std::shared_ptr<tiledb_kv_schema_t>(schema, kv_deleter_);
   }
 
   MapSchema(const MapSchema&) = default;
@@ -325,6 +361,9 @@ class MapSchema : public Schema {
 
   /** The pointer to the C TileDB KV schema object. */
   std::shared_ptr<tiledb_kv_schema_t> schema_;
+
+  /** KV-specific deleter wrapper. */
+  impl::KVDeleter kv_deleter_;
 };
 
 /* ********************************* */

--- a/tiledb/sm/cpp_api/tiledb
+++ b/tiledb/sm/cpp_api/tiledb
@@ -47,8 +47,6 @@
 #include "filter.h"
 #include "filter_list.h"
 #include "group.h"
-#include "map.h"
-#include "map_schema.h"
 #include "object.h"
 #include "object_iter.h"
 #include "query.h"

